File: DotNet Project V2.0/backend-dotnet/src/Emart.Domain/Purchase/PurchaseModeRegistry.cs
namespace Emart.Domain.Purchase;

/// <summary>
/// Factory/registry for <see cref="IPurchaseModeStrategy"/>. DI hands in every strategy
/// implementation it can find, and this indexes them by mode.
///
/// A mode without a strategy, or two strategies claiming the same mode, fails here - callers
/// (Program.cs) resolve this eagerly at startup so that failure happens at STARTUP rather than
/// at the till, mirroring the Spring @Component constructor-time check.
/// </summary>
public sealed class PurchaseModeRegistry
{
    private readonly Dictionary<PurchaseMode, IPurchaseModeStrategy> _strategies = new();

    public PurchaseModeRegistry(IEnumerable<IPurchaseModeStrategy> discovered)
    {
        foreach (var strategy in discovered)
        {
            if (_strategies.TryGetValue(strategy.Mode(), out var clash))
            {
                throw new InvalidOperationException(
                    $"Two purchase-mode strategies claim mode {strategy.Mode()}: " +
                    $"{clash.GetType().FullName} and {strategy.GetType().FullName}");
            }

            _strategies[strategy.Mode()] = strategy;
        }

        foreach (PurchaseMode mode in Enum.GetValues<PurchaseMode>())
        {
            if (!_strategies.ContainsKey(mode))
            {
                throw new InvalidOperationException(
                    $"No IPurchaseModeStrategy registered for purchase mode {mode} " +
                    $"(Mode {mode.ModeNumber()}). Every mode must have exactly one.");
            }
        }
    }

    public IPurchaseModeStrategy ForMode(PurchaseMode? mode)
    {
        var resolved = mode ?? PurchaseMode.CASH_ONLY;

        if (!_strategies.TryGetValue(resolved, out var strategy))
        {
            // Unreachable given the constructor check, but a clear failure beats a
            // NullReferenceException in pricing.
            throw new InvalidOperationException($"No IPurchaseModeStrategy registered for {resolved}");
        }

        return strategy;
    }
}

File: Java Project  V1.7/backend-spring/src/main/java/com/emart/service/purchase/PurchaseModeRegistry.java
package com.emart.service.purchase;

import java.util.EnumMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Component;

/**
 * Factory/registry for {@link PurchaseModeStrategy}.
 *
 * Spring hands in every strategy bean it can find, and this indexes
 * them by mode. Consequences:
 *
 *   - adding a purchase mode = adding one enum value + one @Component;
 *     nothing here or in PurchaseDecisionEngine changes;
 *   - a mode without a strategy, or two strategies claiming the same
 *     mode, fails at STARTUP rather than at the till.
 */
@Component
public class PurchaseModeRegistry {

    private final Map<PurchaseMode, PurchaseModeStrategy> strategies =
            new EnumMap<>(PurchaseMode.class);


    public PurchaseModeRegistry(List<PurchaseModeStrategy> discovered) {

        for (PurchaseModeStrategy strategy : discovered) {

            PurchaseModeStrategy clash =
                    strategies.put(strategy.mode(), strategy);

            if (clash != null) {
                throw new IllegalStateException(
                        "Two purchase-mode strategies claim mode "
                                + strategy.mode() + ": "
                                + clash.getClass().getName() + " and "
                                + strategy.getClass().getName());
            }
        }

        for (PurchaseMode mode : PurchaseMode.values()) {
            if (!strategies.containsKey(mode)) {
                throw new IllegalStateException(
                        "No PurchaseModeStrategy registered for purchase mode "
                                + mode + " (Mode " + mode.getModeNumber()
                                + "). Every mode must have exactly one.");
            }
        }
    }


    public PurchaseModeStrategy forMode(PurchaseMode mode) {

        PurchaseModeStrategy strategy =
                strategies.get(mode == null ? PurchaseMode.CASH_ONLY : mode);

        if (strategy == null) {
            // Unreachable given the constructor check, but a clear
            // failure beats a NullPointerException in pricing.
            throw new IllegalStateException(
                    "No PurchaseModeStrategy registered for " + mode);
        }

        return strategy;
    }
}
